### ISSUE TITLE
GribApiException is only trown in debug mode

### ISSUE THREAD

**Issue by jnyrup** [0x1mason/GribApi.NET#61]
I have a grib2 file which jasper seems to choke on while decoding based on getting "cannot get marker segment" in the console.
Using GribApi.Net master compiled in release mode throws an uncatchable `AccessViolationException` while when compiled in debug mode a `GribApiException` with "Decoding is invalid" is thrown.

```
using (var gribFile = new GribFile(fileName))
{
	foreach (var message in gribFile)
	{
		foreach (var value in message.GeoSpatialValues)
		{
		}
	}
}
```

**Comment by IdahoSixString** [comment]
See Issue #50 on how to handle this. Essentially you just need to create a DTO and transfer all the data out of the GribFile object. Unfortunately, the original developer implemented IEnumerable which is lazily evaluated. Which then when the memory pointers try to go back to read the unmanaged memory it becomes an AccessViolation. Therefore, copy all the unmanaged data into a DTO and your problem will go away.

**Comment by jnyrup** [comment]
@IdahoSixString Thanks for your comment, but that doesn't seem to be the problem here.
I tried out your suggestion, but it still gives me an `AccessViolationException` when calling `gribMessage.GeoSpatialValues.ToList();` when running a release build of Grib.Api.NET.

I dug a bit deeper and the sinner seems to be `GribApiProxyPINVOKE.GribGetErrorMessage(int jarg1)`. When commenting out the call to it in `GribApiProxy.GribGetErrorMessage` a `GribApiException` is correctly thrown.

**Comment by IdahoSixString** [comment]
@jnyrup Interesting a different issue.

**Comment by jnyrup** [comment]
If somebody gets to dig further into this before me, the problems seems to be related to marshalling of the error message string from unmanaged to managed memory.

https://stackoverflow.com/questions/6300093/why-cant-i-return-a-char-string-from-c-to-c-sharp-in-a-release-build
https://blogs.msdn.microsoft.com/dotnet/2014/06/13/net-native-deep-dive-debugging-into-interop-code/

**Comment by 0x1mason** [comment]
@jnyrup Would you be willing to debug this and contribute a PR?

**Comment by jnyrup** [comment]
@0x1mason I did try to do so, but with no immediate luck.
But at some point I will need to do so, to guard us from this uncatchable exceptions.

**Comment by 0x1mason** [comment]
@IdahoSixString I believe this is the same issue as #50, since they are both caused by the logging proc.

**Comment by 0x1mason** [comment]
@IdahoSixString Can you explain how these problems are related to the IEnumerable implementation? `grib_api` uses the very same pattern of lazy evaluation.

**Comment by IdahoSixString** [comment]
@0x1mason jnyrup had already tried my suggestion which it did not successfully solve his issue.

Therefore, I am not sure if this is a problem in relation to IEnumerable as the other issue.

**Comment by 0x1mason** [comment]
This should be resolved in 1.0.0-beta2
